### PR TITLE
DEVX-2487:Add streams error handling with StreamsUncaughtExceptionHandler

### PR DESCRIPTION
**Pull request by bbejeck**
### Description
Adds tutorial for `StreamsUncaughtExceptionHandler` new in AK 2.8

**_NOTE_**: We don't want to merge this PR until AK 2.8.0 is available in Maven Central

### Checklist of tasks required before merging
- [x] Remove local repo - https://github.com/confluentinc/kafka-tutorials/pull/741#discussion_r585137600
- [x] Uncomment lines in `build.gradle` for maven deps - https://github.com/confluentinc/kafka-tutorials/pull/741#discussion_r585138329
- [x] Delete step for creating local folders for KS snapshot jars - https://github.com/confluentinc/kafka-tutorials/pull/741#discussion_r585140095
- [x] Delete line in Makefile copying over jar files for test harness run - https://github.com/confluentinc/kafka-tutorials/pull/741#discussion_r585140651
- [x] Rename title (remove [DO NOT MERGE])

### Staging Docs

http://kafka-tutorials-staging.s3-website-us-west-2.amazonaws.com/DEVX-2487_add_streams_error_handling/
http://kafka-tutorials-staging.s3-website-us-west-2.amazonaws.com/DEVX-2487_add_streams_error_handling/error-handling/kstreams.html

### CODE REVIEW

File: _includes/tutorials/error-handling/kstreams/code/Makefile
Comment (bbejeck) [inline _includes/tutorials/error-handling/kstreams/code/Makefile:11]: This line needs to be deleted before merging

File: _includes/tutorials/error-handling/kstreams/code/build.gradle
Comment (bbejeck) [inline _includes/tutorials/error-handling/kstreams/code/build.gradle:41]: the `files` for the local repo needs to be deleted before emerging, but are needed as 2.8 isn't in Maven Central yet.

Lines above in comments need to be un-commented before merging

File: _includes/tutorials/error-handling/kstreams/markup/dev/build-uberjar.adoc
Comment (bbejeck) [inline _includes/tutorials/error-handling/kstreams/markup/dev/build-uberjar.adoc:10]: This step is here for reviewers who want to run the tutorial locally.  This step is needed to get the necessary jar files available for the build.  This will not be needed once AK 2.8 jars are available in Maven Central.

These lines need to be deleted for merging.

File: _includes/tutorials/error-handling/kstreams/code/Makefile
Comment (bbejeck) [inline _includes/tutorials/error-handling/kstreams/code/Makefile:5]: This line needs to be deleted before merging

**Comment by bbejeck** [comment]
Test for this PR passed https://confluentinc.semaphoreci.com/jobs/3e04ee77-6669-423c-80bb-5b0a41bacaaa

**Comment by bbejeck** [comment]
ping @gAmUssA, @awalther28, @ybyzek, and @rspurgeon for review

File: _includes/tutorials/error-handling/kstreams/markup/dev/make-exception-handler.adoc
Comment (ybyzek) [inline _includes/tutorials/error-handling/kstreams/markup/dev/make-exception-handler.adoc:54]:
```suggestion
The idea here is that a couple of errors spread out are ok so processing continues.  But a bunch of errors withing a small window of time could indicate a bigger issue, so it's better to shutdown.  While the code doesn't inspect the type of the exception, that's another valid approach as well.
```

File: _includes/tutorials/error-handling/kstreams/markup/dev/make-exception-handler.adoc
Comment (ybyzek) [inline _includes/tutorials/error-handling/kstreams/markup/dev/make-exception-handler.adoc:56]:
```suggestion
The above code is just an example of what you could do and definitely not tested in a production setting.  The main point here is while it's a good idea to keep processing with a small number of errors, it's not a good idea to continually replace the thread with sustained errors.  It's better to have some "guard rails" in place to make sure your application is robust, but won't continue on when it shouldn't.
```

File: _includes/tutorials/error-handling/kstreams/markup/dev/make-exception-handler.adoc
Comment (ybyzek) [inline _includes/tutorials/error-handling/kstreams/markup/dev/make-exception-handler.adoc:64]:
```suggestion
You'll add the `StreamsUncaughtExceptionHandler` to your Kafka Streams application in the next step.
```

File: _includes/tutorials/error-handling/kstreams/markup/dev/make-topology.adoc
Comment (ybyzek) [inline _includes/tutorials/error-handling/kstreams/markup/dev/make-topology.adoc:34]:
```suggestion
<1> Simulating an error depending on the value of a counter (which gets incremented with every record)
```

File: _includes/tutorials/error-handling/kstreams/markup/dev/make-topology.adoc
Comment (ybyzek) [inline _includes/tutorials/error-handling/kstreams/markup/dev/make-topology.adoc:38]: For your consideration, move this sentence up to follow "Here is the code we’ll use to drive our tutorial"

```suggestion
This code ensures that the rate of errors (3 within a 1 hour window) meets the criteria for shutting down the application.
```

File: _includes/tutorials/error-handling/kstreams/markup/dev/run-dev-app.adoc
Comment (ybyzek) [inline _includes/tutorials/error-handling/kstreams/markup/dev/run-dev-app.adoc:10]:
```suggestion
Now run the following program, but watch the logs in the console and let the application run for a few seconds.
You should observe it shutting down.
```

Comment (ybyzek) [inline _includes/tutorials/error-handling/kstreams/markup/dev/run-dev-app.adoc:10]: Can you add a sample output from the logs that shows it shutting down?

File: _includes/tutorials/error-handling/kstreams/markup/dev/run-consumer.adoc
Comment (ybyzek) [inline _includes/tutorials/error-handling/kstreams/markup/dev/run-consumer.adoc:34]:
```suggestion
You'll notice there are some duplicated values in the output.  This duplication is to be expected, as the streams application is running with the default processing mode of `AT_LEAST_ONCE`. Duplicate values is one thing to consider when using `REPLACE_THREAD` with the `StreamsUncaughtExceptionHander`, since this is analogous to using retries with the `KafkaProducer`.  If you don't want duplicate values, you should consider running with the processing mode of `EXACTLY_ONCE`
```

Comment (ybyzek) [inline _includes/tutorials/error-handling/kstreams/markup/dev/run-consumer.adoc:34]: Can you add this configuration parameter into the application?  That way the user can see how it is configured, and optionally change it for testing.

File: _includes/tutorials/error-handling/kstreams/markup/dev/make-exception-handler.adoc
Comment (ybyzek) [inline _includes/tutorials/error-handling/kstreams/markup/dev/make-exception-handler.adoc:13]: @bbejeck this might be a good addition here since duplicates are a big deal for some use cases.  Please note the placeholder `<insert configuration parameter name>`

```suggestion
* REPLACE_THREAD - Replaces the thread receiving the exception and processing continues with the same number of configured threads. (Note: this can result in duplicate records depending on the application's processing mode <insert configuration parameter name>)
```

File: _includes/tutorials/error-handling/kstreams/markup/dev/run-consumer.adoc
Comment (ybyzek) [inline _includes/tutorials/error-handling/kstreams/markup/dev/run-consumer.adoc:9]:
```suggestion
Let's now run the `kafka-console-consumer` to confirm the output:
```

File: _includes/tutorials/error-handling/kstreams/markup/dev/make-exception-handler.adoc
Comment (ybyzek) [inline _includes/tutorials/error-handling/kstreams/markup/dev/make-exception-handler.adoc:49]:
```suggestion
<1> Checking if the current number of failures equals or exceeds
```

File: _includes/tutorials/error-handling/kstreams/markup/dev/run-consumer.adoc
Comment (ybyzek) [inline _includes/tutorials/error-handling/kstreams/markup/dev/run-consumer.adoc:15]:
```suggestion
Your results should look something like this:
```

File: _includes/tutorials/error-handling/kstreams/markup/dev/make-exception-handler.adoc
Comment (ybyzek) [inline _includes/tutorials/error-handling/kstreams/markup/dev/make-exception-handler.adoc:15]: How does it do this if the other applications are on other systems?

**Review by ybyzek — COMMENTED** [review]
@bbejeck made a first pass-thru, looking good!  I did code/text review only; I did not run the code (ran out of time) but wanted to get some feedback in

**Review by ybyzek — COMMENTED** [review]
@bbejeck not sure if I just missed it, but it may be helpful to include a link somewhere to the KIP and name it.  https://cwiki.apache.org/confluence/display/KAFKA/KIP-671%3A+Introduce+Kafka+Streams+Specific+Uncaught+Exception+Handler

File: _includes/tutorials/error-handling/kstreams/markup/answer/error-handling-answer.adoc
Comment (ybyzek) [inline _includes/tutorials/error-handling/kstreams/markup/answer/error-handling-answer.adoc:1]:
```suggestion
To handle uncaught exceptions, use the  https://kafka.apache.org/28/javadoc/org/apache/kafka/streams/KafkaStreams.html#setUncaughtExceptionHandler-org.apache.kafka.streams.errors.StreamsUncaughtExceptionHandler-[KafkaStreams.setUncaughtExceptionHandler] method, and configure one of three valid actions: `REPLACE_THREAD`, `SHUTDOWN_CLIENT`, or `SHUTDOWN_APPLICATION`.
```

File: _data/tutorials.yaml
Comment (ybyzek) [inline _data/tutorials.yaml:678]: @bbejeck is there a chance a ksqlDB app would be written for this?  If so, would it make sense to generalize the question with the `streams` word?

File: index.html
Comment (ybyzek) [inline index.html:59]: @bbejeck note that there is an existing use case for "Handle deserialization errors" (https://kafka-tutorials.confluent.io/handling-deserialization-errors/ksql.html).  It would be good to think about how to name each of them so that they are clear & distinct and won't be confused with eachother.  (e.g. right now, "Error handling" sounds like a superset that could include "Handle deserialization errors" -- maybe this isn't ideal?)

File: _includes/tutorials/error-handling/kstreams/markup/answer/error-handling-answer.adoc
Comment (bbejeck) [reply _includes/tutorials/error-handling/kstreams/markup/answer/error-handling-answer.adoc:1]: ack added something very similar

File: _includes/tutorials/error-handling/kstreams/markup/dev/run-dev-app.adoc
Comment (bbejeck) [reply _includes/tutorials/error-handling/kstreams/markup/dev/run-dev-app.adoc:10]: ack

**Comment by bbejeck** [comment]
@ybyzek thanks for the thorough review, I believe I've addressed all comments.

File: _includes/tutorials/error-handling/kstreams/code/docker-compose.yml
Comment (bbejeck) [reply _includes/tutorials/error-handling/kstreams/code/docker-compose.yml:6]: doh! missed that

Comment (bbejeck) [reply _includes/tutorials/error-handling/kstreams/code/docker-compose.yml]: done

**Comment by bbejeck** [comment]
The test for this PR passed https://confluentinc.semaphoreci.com/jobs/dc44b1cb-f636-4000-a790-180e4905154b

**Comment by bbejeck** [comment]
ping @ybyzek all comments addressed ready for another review

File: _includes/tutorials/error-handling/kstreams/code/src/main/java/io/confluent/developer/StreamsUncaughtExceptionHandling.java
Comment (rubieHess) [inline _includes/tutorials/error-handling/kstreams/code/src/main/java/io/confluent/developer/StreamsUncaughtExceptionHandling.java:149]: I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [FS: Format string should use %n rather than n (VA_FORMAT_STRING_USES_NEWLINE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#fs-format-string-should-use-n-rather-than-n-va-format-string-uses-newline).
This format string includes a newline character (\n). In format strings, it is generally  preferable to use %n, which will produce the platform-specific line separator.

**Review by ybyzek — APPROVED** [review]
@bbejeck @awalther28 since I already made a pass (code/text review only, I did not run the KT) would either of you have concerns if I yielded my second review to @awalther28 ?

File: _includes/tutorials/error-handling/kstreams/code/src/main/java/io/confluent/developer/StreamsUncaughtExceptionHandling.java
Comment (bbejeck) [reply _includes/tutorials/error-handling/kstreams/code/src/main/java/io/confluent/developer/StreamsUncaughtExceptionHandling.java]: @rubieHess good catch, I've fixed this.

Comment (rubieHess) [reply _includes/tutorials/error-handling/kstreams/code/src/main/java/io/confluent/developer/StreamsUncaughtExceptionHandling.java]: Happy to help:)

**Comment by bbejeck** [comment]
@ybyzek I've made all the changes necessary since AK 2.8 is released now

**Comment by ybyzek** [comment]
👍 beauty!  And thanks for updating the checklist in the Description.  Is the only thing blocking this merge the Semaphore issue?

**Comment by bbejeck** [comment]
The relevant test for this PR passed https://confluentinc.semaphoreci.com/jobs/caf6b5c8-f7e0-4aa2-9934-643f78506d0f 

There seems to be an issue with the `KStreams emit a final result from a time window` but I'll handle that in a separate PR

**Pull request merged**
